### PR TITLE
[Carbon/C++ interop] Add more tests for `int` function params support

### PR DESCRIPTION
**Pull request by ivanaivanovska**
Added tests for `signed int`, `signed`, `std::int32_t` and `const int&`.
This is an addition to PR #5197.

Part of #5064

### CODE REVIEW

File: toolchain/check/testdata/interop/cpp/function_decl.carbon [hunk 1]
```diff
@@ -45,9 +45,55 @@ fn F() {
   Cpp.foo_float();
 }
 
+// --- signed_int_param_function_decl.h
```
Comment (bricknerb) [inline toolchain/check/testdata/interop/cpp/function_decl.carbon:62]: I suggest putting this test after the `int` test since `int` is more common.

File: toolchain/check/testdata/interop/cpp/function_decl.carbon [hunk 2]
```diff
@@ -45,9 +45,55 @@ fn F() {
...
+import Cpp library "signed_param_function_decl.h";
+
+fn F() {
+  Cpp.foo(1);
+}
+
+// --- int32_t_param_function_decl.h
+
+namespace std {
+  typedef signed int int32_t;
```
Comment (bricknerb) [inline toolchain/check/testdata/interop/cpp/function_decl.carbon:94]: Consider pointing to how `int32_t` is actually defined in the standard library.

File: toolchain/check/testdata/interop/cpp/function_decl.carbon [hunk 3]
```diff
@@ -134,6 +180,28 @@ fn F() {
 }
 
 
+// --- int_const_ref_param_function_decl.h
```
Comment (bricknerb) [inline toolchain/check/testdata/interop/cpp/function_decl.carbon]: What about `const int` and `const int*`?

**Review by bricknerb — APPROVED** [review]

File: toolchain/check/testdata/interop/cpp/function_decl.carbon [hunk 4]
```diff
@@ -45,9 +45,55 @@ fn F() {
...
+auto foo(signed int a) -> void;
+
+// --- import_signed_int_param_function_decl.carbon
+
+library "[[@TEST_NAME]]";
+
+import Cpp library "signed_int_param_function_decl.h";
+
+fn F() {
+  Cpp.foo(1);
```
Comment (danakj) [inline toolchain/check/testdata/interop/cpp/function_decl.carbon:73]: Should we test the boundary conditions too for these tests? For this case, we could test passing INT_MAX and INT_MIN, and a fail test that passes values outside that range?

File: toolchain/check/testdata/interop/cpp/function_decl.carbon [hunk 5]
```diff
@@ -134,6 +180,28 @@ fn F() {
 }
 
 
+// --- int_const_ref_param_function_decl.h
+
+auto foo(const int& a) -> int;
+
+// --- fail_int_const_ref_param_function_decl.carbon
```
Comment (danakj) [inline toolchain/check/testdata/interop/cpp/function_decl.carbon]: If this is meant to pass in the future, then
```suggestion
// --- fail_todo_int_const_ref_param_function_decl.carbon
```

File: toolchain/check/testdata/interop/cpp/function_decl.carbon (hunk 1 above)
Comment (ivanaivanovska) [reply toolchain/check/testdata/interop/cpp/function_decl.carbon:62]: Done.

File: toolchain/check/testdata/interop/cpp/function_decl.carbon (hunk 4 above)
Comment (ivanaivanovska) [reply toolchain/check/testdata/interop/cpp/function_decl.carbon:73]: sg, done.

File: toolchain/check/testdata/interop/cpp/function_decl.carbon (hunk 2 above)
Comment (ivanaivanovska) [reply toolchain/check/testdata/interop/cpp/function_decl.carbon:94]: added a description, please let me know if that's what you mean.

File: toolchain/check/testdata/interop/cpp/function_decl.carbon (hunk 3 above)
Comment (ivanaivanovska) [reply toolchain/check/testdata/interop/cpp/function_decl.carbon]: Added a test for `const int*`. `const int` is not recommended in function declarations, only in definitions, so I skipped that one.

File: toolchain/check/testdata/interop/cpp/function_decl.carbon (hunk 5 above)
Comment (ivanaivanovska) [reply toolchain/check/testdata/interop/cpp/function_decl.carbon]: thx, changed the other tests that are supposed to pass in the future as well.

File: toolchain/check/testdata/interop/cpp/function_decl.carbon (hunk 2 above)
Comment (danakj) [reply toolchain/check/testdata/interop/cpp/function_decl.carbon:94]: I think a link to the libcxx sources that show how it defines int32_t might be most appropriate. We want to increase confidence that this is testing what actual `std` will be like

File: toolchain/check/testdata/interop/cpp/function_decl.carbon (hunk 3 above)
Comment (danakj) [reply toolchain/check/testdata/interop/cpp/function_decl.carbon]: Recommended is not the high bit here, I think. If there's valid c++ code that does it, we want to be able to interop with it

File: toolchain/check/testdata/interop/cpp/function_decl.carbon (hunk 2 above)
Comment (ivanaivanovska) [reply toolchain/check/testdata/interop/cpp/function_decl.carbon:94]: sg, done.

File: toolchain/check/testdata/interop/cpp/function_decl.carbon (hunk 3 above)
Comment (ivanaivanovska) [reply toolchain/check/testdata/interop/cpp/function_decl.carbon]: ok, added a test for `const int` as well.

**Review by danakj — APPROVED** [review]
LGTM

File: toolchain/check/testdata/interop/cpp/function_decl.carbon [hunk 6]
```diff
@@ -28,32 +28,79 @@ fn F() {
...
+import Cpp library "signed_param_function_decl.h";
+
+fn F() {
+  Cpp.foo(1);
+}
+
+// --- int32_t_param_function_decl.h
+
+namespace std {
+  // Mimicing Clang's definition for int32_t: https://github.com/llvm/llvm-project/blob/fd800487382e2ee3944493d58a961b6f48290243/clang/lib/Headers/stdint.h#L197
```
Comment (bricknerb) [inline toolchain/check/testdata/interop/cpp/function_decl.carbon]:
```suggestion
  // Mimicking Clang's definition for int32_t: https://github.com/llvm/llvm-project/blob/fd800487382e2ee3944493d58a961b6f48290243/clang/lib/Headers/stdint.h#L197
```

File: toolchain/check/testdata/interop/cpp/function_decl.carbon (hunk 2 above)
Comment (bricknerb) [reply toolchain/check/testdata/interop/cpp/function_decl.carbon:94]: FWIW, this still points to `clang` and not to `libcxx`.
I'm not sure what is @danakj intention was though. In libcxx I could find [this](https://github.com/llvm/llvm-project/blob/e3ee6bbd384ef4c583b9f7bca4253ae0fba90a70/libcxx/include/cstdint#L160), but I'm not sure this is helpful.

File: toolchain/check/testdata/interop/cpp/function_decl.carbon [hunk 7]
```diff
@@ -133,66 +246,123 @@ fn F() {
...
+  // CHECK:STDERR: fail_todo_int_pointer_param_function_decl.carbon:[[@LINE+4]]:3: note: in `Cpp` name lookup for `foo` [InCppNameLookup]
   // CHECK:STDERR:   Cpp.foo(&a);
   // CHECK:STDERR:   ^~~~~~~
   // CHECK:STDERR:
   Cpp.foo(&a);
 }
 
+// --- int_const_param_function_decl.h
+
+auto foo(const int a) -> void;
```
Comment (bricknerb) [inline toolchain/check/testdata/interop/cpp/function_decl.carbon]: Order wise, I would put `const int` before `const int*` and `const int&`.

File: toolchain/check/testdata/interop/cpp/function_decl.carbon [hunk 8]
```diff
@@ -108,21 +155,87 @@ fn F() {
...
   // CHECK:STDERR:   ^~~~~~~~~
-  // CHECK:STDERR: fail_int_default_param_parameterless_call.carbon: note: calling function declared here [InCallToEntity]
+  // CHECK:STDERR: fail_todo_int_default_param_parameterless_call.carbon: note: calling function declared here [InCallToEntity]
   // CHECK:STDERR:
   Cpp.foo();
 }
 
+// --- int_max_param_function_decl.h
+
+auto foo(int a) -> void;
```
Comment (bricknerb) [inline toolchain/check/testdata/interop/cpp/function_decl.carbon:175]: This exact function is defined multiple times in several header files.
We might want to share the same header here and put all the tests that use it together.

File: toolchain/check/testdata/interop/cpp/function_decl.carbon [hunk 9]
```diff
@@ -108,21 +155,87 @@ fn F() {
...
+
+// --- import_int_min_param_function_decl.carbon
+
+library "[[@TEST_NAME]]";
+
+import Cpp library "int_min_param_function_decl.h";
+
+fn F() {
+  Cpp.foo(-0x8000_0000);
```
Comment (bricknerb) [inline toolchain/check/testdata/interop/cpp/function_decl.carbon:217]: I think writing `-0x8000_0000` is a bit confusing since in C++, you could write `0x8000_0000` (without the minus) to calculate INT_MIN.
You can also consider writing `-2147483648` explicitly.

**Review by bricknerb — APPROVED** [review]

File: toolchain/check/testdata/interop/cpp/function_decl.carbon (hunk 2 above)
Comment (danakj) [reply toolchain/check/testdata/interop/cpp/function_decl.carbon:94]: Oh, thanks, yes that's the link I was expecting. But yeah it's just a forward of the root one. Here's the definition on gnu/linux: https://codebrowser.dev/glibc/glibc/posix/bits/types.h.html#__int32_t

Comment (ivanaivanovska) [reply toolchain/check/testdata/interop/cpp/function_decl.carbon:94]: I updated now both the description and the link, hopefully it's now correct.

File: toolchain/check/testdata/interop/cpp/function_decl.carbon (hunk 6 above)
Comment (ivanaivanovska) [reply toolchain/check/testdata/interop/cpp/function_decl.carbon]: Done.

File: toolchain/check/testdata/interop/cpp/function_decl.carbon (hunk 7 above)
Comment (ivanaivanovska) [reply toolchain/check/testdata/interop/cpp/function_decl.carbon]: sg, done.

File: toolchain/check/testdata/interop/cpp/function_decl.carbon (hunk 9 above)
Comment (danakj) [reply toolchain/check/testdata/interop/cpp/function_decl.carbon:217]: FWIW I do prefer it as written. Carbon literals have different behaviour and this is a test of Carbon code (calling C++), not a test of C++ code.

File: toolchain/check/testdata/interop/cpp/function_decl.carbon [hunk 10]
```diff
@@ -90,8 +90,8 @@ fn F() {
 // --- int32_t_param_function_decl.h
 
 namespace std {
-  // Mimicing Clang's definition for int32_t: https://github.com/llvm/llvm-project/blob/fd800487382e2ee3944493d58a961b6f48290243/clang/lib/Headers/stdint.h#L197
-  typedef int int32_t;
+  // Mimicking libc++ definition for int32_t: https://codebrowser.dev/glibc/glibc/posix/bits/types.h.html#__int32_t
```
Comment (danakj) [inline toolchain/check/testdata/interop/cpp/function_decl.carbon]: This should now say glibc not libc++ (sorry for all this churn..)

Comment (ivanaivanovska) [reply toolchain/check/testdata/interop/cpp/function_decl.carbon]: np, fixed.

File: toolchain/check/testdata/interop/cpp/function_decl.carbon (hunk 9 above)
Comment (ivanaivanovska) [reply toolchain/check/testdata/interop/cpp/function_decl.carbon:217]: sg, keeping the code as it was (`-0x8000_0000`).

File: toolchain/check/testdata/interop/cpp/function_decl.carbon (hunk 8 above)
Comment (ivanaivanovska) [reply toolchain/check/testdata/interop/cpp/function_decl.carbon:175]: Agree. But let's do this refactoring in a follow-up and fix the remaining tests as well.

**Comment by ivanaivanovska** [comment]
@danakj @bricknerb all comments should be addressed now. If it looks fine, can you please help merge the PR? Thanks.

**Comment by danakj** [comment]
looks like theres a test failure. you might need to run autoupdate?

**Comment by bricknerb** [comment]
> looks like theres a test failure. you might need to run autoupdate?

Yes, I think we needed to merge and autoupdate, I believe it's done now.
